File: inst/app/serve.py
#!/usr/bin/env python3
"""Legacy static-file helper. RGrid-local is normally served from R with openRGrid()."""

from argparse import ArgumentParser
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import os


class Handler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        self.send_header("Cross-Origin-Resource-Policy", "cross-origin")
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()


def main():
    parser = ArgumentParser()
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", default=8080, type=int)
    args = parser.parse_args()
    os.chdir(Path(__file__).resolve().parent)
    server = ThreadingHTTPServer((args.host, args.port), Handler)
    print(f"RGrid: http://{args.host}:{args.port}")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopping.")
    finally:
        server.server_close()


if __name__ == "__main__":
    main()

File: inst/app/serve.R
# Run with: Rscript serve.R
# Requires install.packages("httpuv") once in the system R installation.

library(httpuv)

args <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", args[grep("^--file=", args)][1])
root <- dirname(normalizePath(file_arg, mustWork = TRUE))
setwd(root)
cat("RGrid: http://127.0.0.1:8080\n")

runServer(
  host = "127.0.0.1",
  port = 8080,
  app = list(
    staticPaths = list(
      "/" = staticPath(
        ".",
        headers = list(
          "Cross-Origin-Opener-Policy" = "same-origin",
          "Cross-Origin-Embedder-Policy" = "require-corp",
          "Cross-Origin-Resource-Policy" = "cross-origin",
          "Cache-Control" = "no-cache"
        )
      )
    )
  )
)
